File: 2025/day07.py
with open("2025/input/day07") as f:
    data = f.read().splitlines()

    diagram = [list(row) for row in data]
    potential = [[0 for _ in range(len(diagram[0]))] for _ in range(len(diagram))]

    for j in range(len(diagram[0])):
        if diagram[0][j] == "S":
            diagram[1][j] = "|"
            potential[1][j] = 1
    for i in range(1, len(diagram) - 1):
        for j in range(len(diagram[i])):
            if diagram[i][j] == "^":
                diagram[i][j - 1] = "|"
                potential[i][j - 1] += potential[i - 1][j]
                diagram[i][j + 1] = "|"
                potential[i][j + 1] += potential[i - 1][j]
        for j in range(len(diagram[i])):
            if diagram[i][j] == "|" and diagram[i + 1][j] == ".":
                diagram[i + 1][j] = "|"
                potential[i + 1][j] += potential[i][j]

    partie1 = 0

    for i in range(len(diagram) - 1):
        for j in range(len(diagram[i])):
            if diagram[i + 1][j] == "^" and diagram[i][j] == "|":
                partie1 += 1

    print("Partie 1:", partie1)

    partie2 = sum(x for x in potential[-1])

    print("Partie 2:", partie2)

File: 2025/rust/src/aoc2025/day07.rs
use rdcl_aoc_helpers::input::WithReadLines;
use std::fs::File;

pub fn day07(path: &String) {
    let mut diagram: Vec<Vec<char>> = File::open(path).read_lines::<String>(1).map(|x| x.chars().collect()).collect();
    let mut potential: Vec<Vec<u64>> = Vec::new();

    for _i in 0..diagram.len() {
        potential.push(vec![0; diagram[0].len()]);
    }

    for i in 0..diagram[0].len() {
        if diagram[0][i] == 'S' {
            diagram[1][i] = '|';
            potential[1][i] = 1; 
        }
    }

    for i in 1..diagram.len() -1 {
        for j in 0..diagram[0].len() {
            if diagram[i][j] == '^'{
                diagram[i][j - 1] = '|';
                potential[i][j - 1] += potential[i - 1][j];
                diagram[i][j + 1] = '|';
                potential[i][j + 1] += potential[i - 1][j];
            }
        }
        for j in 0..diagram[0].len() {
            if diagram[i][j] == '|' && diagram[i + 1][j] == '.'{
                diagram[i + 1][j] = '|';
                potential[i + 1][j] += potential[i][j];
            }
        }
    }

    let mut partie1 = 0;

    for i in 0..diagram.len() - 1 {
        for j in 0..diagram[0].len() {
            if diagram[i+1][j] == '^' && diagram[i][j] == '|' {
                partie1 += 1;
            }
        }
    }

    let partie2:u64 = potential[diagram.len() - 1].iter().sum();

    println!("Partie 1: {}", partie1);
    println!("Partie 2: {}", partie2);
}